tests/test.sh

```
@@ -24,79 +23,100 @@ ACTION="$7"
-            echo "File $file differs from the reference."
-         fi
-      fi
-   fi
-done
-return $status
+function diff_files {
+  local return_status=0
+  local error_code
+  local ref_file
+  local test_file
+  # Compare test results with all existing reference files
+  local reference_files=$(ls *.ref)
+  if [[ -z $reference_files ]];then
+    echo "ERROR: No reference files were found"
+    return 1
+  fi
+
+  for ref_file in $reference_files
+  do
+    test_file=$(basename $ref_file .ref)
+    if [[ ! -f $test_file ]];then
+      # The output file does not exist.
+      # Something went seriously wrong, ABIN probably crashed prematurely.
+      # No need for further checks, exit NOW.
+      echo "ERROR: Could not find output file \"$test_file\""
+      return 1
+    fi
+
+    error_code=0
+    diff -q $test_file $ref_file || error_code=$?
+    if [[ $error_code -ne 0 ]];then
+       # The reference file is different, but maybe it's just numerical noise?
+       error_code=0
+       diff -y -W 500  $test_file $ref_file | egrep -e '|' -e '<' -e '>' > $test_file.diff
+
+       ../numdiff.py $test_file.diff || error_code=$?
+
+       if [[ $error_code -ne 0 ]];then
+          # The changes were bigger that the thresholds specified in numdiff.py
+          return_status=$error_code
+          echo "File $test_file differs from the reference."
+       fi
+    fi
+  done
+  return $return_status
 }
 
+# Update already existing reference files.
+# Called by `make makeref TEST=TEST_FOLDER`
+# If you're creating a completely new test,
+# you need to create the reference files manually.
 function makeref {
-local files
-local f
-echo "Making new reference files."
-files=$(ls *.ref)
-for f in $files 
-do
-   file=$(basename $f .ref)
-   if [[ -f $file.ref ]];then
-      mv $file $file.ref
-   fi
-done
+  local ref_file
+  local test_file
+  local reference_files=$(ls *.ref)
+  echo "Making new reference files."
+  if [[ -z $reference_files ]];then
```

Here again, we fail loudly if no pre-existing reference files were found. And that is good.